### DISCUSSION TITLE
Skipping the interpreter to an arbitrary state

### DISCUSSION THREAD

**Discussion by DeadWisdom** [statelyai/xstate#2214] [Q&A]
Currently, there doesn't seem to be a way to skip the interpreter to a given state.

My use-case is that I have a SPA wizard that I am modeling as states.  Each step is a state. When the user presses next, the state transitions to a new one, and the URL updates in response via history.pushState().

Now when the user presses back or is linked to a different place in the flow, I want to move right to that state. In reality, each state can be transitioned directly to another state, but that doesn't seem constructive to model.

I am using it in preact like:

```typescript
    const [state, send, service] = useMachine(MyStateMachine);
```

And when the page changes, my hack is thus:

```typescript
    const newState = MyStateMachine.transition(newStep, 'GOTO');
    (service as any).update(newState, newState.event);
```

What's the best way to do this? Not sure if there are issues with my solution as `update()` is private.

Thanks

**Comment by ChrisShank** [comment]
The way that I have been thinking about routing with `xstate` is by really embracing the actor model. The browser tab *actually* owns the URL so we should treat it as an actor whom we can message with through the History API (i.e send events by `pushState` and receive events by listening to `popstate`). The machine for the multi-step form invokes the "router", which is just sending and receiving events from the History API. Might look something like this:

```ts
const machine = createMachine({
  id: 'form-wizard'
  initial: 'idle',
  states: {
    idle: {
      on: {
        START: { 
          target: 'step1',
          actions: send({ type: 'step', step: 1 }, { to: 'router' })
        }
      }
    },
    step1: {
      on: {
        NEXT: { 
          target: 'step2',
          actions: send({ type: 'step', step: 2 }, { to: 'router' })
        },
        navigateToStep1: undefined, // Dont navigate back if we are already there
      }
    },
    step2: {
      on: {
        NEXT: { 
          target: 'step3',
          actions: send({ type: 'step', step: 3 }, { to: 'router' })
        },
        navigateToStep2: undefined, // Don't navigate back if we are already there
      }
    },
    step3: {
      on: {
        SUBMIT: { 
          target: 'submit',
          actions: send({ type: 'submit' }, { to: 'router' })
        },
        navigateToStep3: undefined, // Don't navigate back if we are already there
      }
    },
    submit: {},
  },
  // Navigation to or from can happen at any point
  on: {
    navigateToStep1: 'step1',
    navigateToStep2: 'step2',
    navigateToStep3: 'step3',
  },
  invoke: { src: 'router' },
}, {
  services: {
    router: () => (sendBack, receive) => {
      // figure out where to go base on the initial URL
      sendBack(...)

      addEventListener('popstate', (e) => {
        // determine what state to send back to the machine based on the new URL
        sendBack(...)
     })

     receive(event => {
       if (event.type === 'step') {
         history.pushState({}, '', `/step/${event.step}`)
       } else if (event.type === 'submit') {
         history.pushState({}, '', `/submit`)
       }
     })
    }
  }
})
```

This is something I am still thinking through so maybe take it with a grain of salt! There might be other ways to accomplish this as well! Here is a [working example](https://codesandbox.io/s/routing-with-xstate-poc-gm9km?file=/src/index.ts) of the same pattern if it helps.
